# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd

path = '/content/drive/MyDrive/data/supermarket_sales.csv'
df = pd.read_csv(path)

df.head()

Mounted at /content/drive


,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [ ]:
df.head(4)


,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


In [ ]:
df.tail(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [ ]:
df.loc[615:619]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
615,509-29-3912,A,Yangon,Member,Female,Sports and travel,38.72,9,17.4240,365.9040,3/20/2019,12:24,Ewallet,348.48,4.761905,17.4240,4.2
616,327-40-9673,B,Mandalay,Member,Male,Sports and travel,72.60,6,21.7800,457.3800,1/13/2019,19:51,Cash,435.60,4.761905,21.7800,6.9
617,840-19-2096,C,Naypyitaw,Member,Male,Electronic accessories,87.91,5,21.9775,461.5275,3/14/2019,18:10,Ewallet,439.55,4.761905,21.9775,4.4
618,828-46-6863,A,Yangon,Member,Male,Food and beverages,98.53,6,29.5590,620.7390,1/23/2019,11:22,Credit card,591.18,4.761905,29.5590,4.0
619,641-96-3695,C,Naypyitaw,Member,Female,Fashion accessories,43.46,6,13.0380,273.7980,2/7/2019,17:55,Ewallet,260.76,4.761905,13.0380,8.5


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [ ]:
df.dtypes

,0
Invoice ID,object
Branch,object
City,object
Customer type,object
Gender,object
Product line,object
Unit price,float64
Quantity,int64
Tax 5%,float64
Total,float64


Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [ ]:
df.describe()

,Unit price,Quantity,Tax 5%,Total,cogs,gross margin percentage,gross income,Rating
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.00000,1.000000e+03,1000.000000,1000.00000
mean,55.672130,5.510000,15.379369,322.966749,307.58738,4.761905e+00,15.379369,6.97270
std,26.494628,2.923431,11.708825,245.885335,234.17651,6.131498e-14,11.708825,1.71858
min,10.080000,1.000000,0.508500,10.678500,10.17000,4.761905e+00,0.508500,4.00000
25%,32.875000,3.000000,5.924875,124.422375,118.49750,4.761905e+00,5.924875,5.50000
50%,55.230000,5.000000,12.088000,253.848000,241.76000,4.761905e+00,12.088000,7.00000
75%,77.935000,8.000000,22.445250,471.350250,448.90500,4.761905e+00,22.445250,8.50000
max,99.960000,10.000000,49.650000,1042.650000,993.00000,4.761905e+00,49.650000,10.00000


In [ ]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.50, 0.75, 0.99])

,gross income
count,1000.000000
mean,15.379369
std,11.708825
min,0.508500
1%,0.913735
25%,5.924875
50%,12.088000
75%,22.445250
99%,45.250750
max,49.650000


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [6]:
df.City.value_counts()

,count
City,
Yangon,340
Mandalay,332
Naypyitaw,328


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [9]:
invoice_counts = df.City.value_counts()
print(invoice_counts / invoice_counts.sum())

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: count, dtype: float64


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [24]:
yangon_mean = df[df.City == 'Yangon']['gross income'].mean()
mandalay_mean = df[df.City == 'Mandalay']['gross income'].mean()

yangon_mean, mandalay_mean


(np.float64(14.874001470588235), np.float64(15.232024096385542))

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

In [24]:
yangon_total=df[df.City =='Yangon']['Total'].mean()
mandalay_total=df[df.City =='Mandalay']['Total'].mean()

yangon_total,mandalay_total



(np.float64(312.35403088235296), np.float64(319.87250602409637))

Можемо побачити, що в місті Mandalay більше і gross income, і привязаний до нього показник Total (загальна сума покупок), що також є важливим показником, тому в аналізі будемо спиратися і на нього. Водночас бачимо, що кількість інвойсів в Mandalay менша - 332 проти 340 в Yangon, тому вищий показник загальних продажів в Mandalay, можна зробити висновок, - досягається за рахунок більших сум чеків.
Тому маємо 3 гіпотези, чим більший Total в місті Mandalay досягається:
1) більшими сумами продажів загалом - тобто більшими середніми чеками, ніж в місті Yangon (наприклад, більшою кількістю товарів Quantity в чеках, комбінацією кількох категорій товарів в одній покупці)
2) більшими цінами на товари в місті Mandalay - вищий Unit price- і як результат, більшими сумами чеків; продаж більш дорогих категорій товарів - що автоматично також підвищує суму середнього чеку
3) більшою часткою крупних чеків (покупок з високою сумою), що зміщує середній Total вгору. Якщо говорити про розподіл чеків Total в даному випадку -то можна сказати про більший правий хвост розподілу (це можна буде побачити також пізніще в значенні персентилів).


In [4]:
yangon_quantity = df[df.City=='Yangon']['Quantity'].mean()
mandalay_quantity = df[df.City=='Mandalay']['Quantity'].mean()

yangon_quantity,mandalay_quantity

(np.float64(5.4676470588235295), np.float64(5.481927710843373))

In [27]:
yangon_unprice= df[df.City=='Yangon']['Unit price'].mean()
mandalay_unprice= df[df.City=='Mandalay']['Unit price'].mean()

yangon_unprice,mandalay_unprice

(np.float64(54.78085294117646), np.float64(55.65927710843374))

Порахуємо кількість товарів в чеках (середню) та середні ціни на товари в містах Mandalay та Yangon для перевірки наших перших двох гіпотез. Як бачимо, в обоз мітсах кількість товарі в чеках в середньому не має значущої різниці і майже ожнакова. Тому першу гіпотезу можемо відкинути.
Тоді як середня ціна на товари трохи вища - і саме вона може давати різницю в чеках в більший бік. Тому другу гвпотезу поки залищаємо.

In [30]:
df[df.City=='Yangon']['Product line'].value_counts()


,count
Product line,
Home and lifestyle,65
Electronic accessories,60
Sports and travel,59
Food and beverages,58
Fashion accessories,51
Health and beauty,47


In [31]:
df[df.City=='Mandalay']['Product line'].value_counts()

,count
Product line,
Fashion accessories,62
Sports and travel,62
Electronic accessories,55
Health and beauty,53
Food and beverages,50
Home and lifestyle,50


8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [23]:
yangon = df[df.City == 'Yangon']['gross margin percentage'].mean()
mandalay = df[df.City == 'Mandalay']['gross margin percentage'].mean()

yangon, mandalay


(np.float64(4.761904762), np.float64(4.761904762))

У всіх всіх інвойсах однакова маржа, власне тому що однакове значення Tax (5%)

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [35]:
df[df.City=='Yangon']['Product line'].value_counts(normalize=True)

,proportion
Product line,
Home and lifestyle,0.191176
Electronic accessories,0.176471
Sports and travel,0.173529
Food and beverages,0.170588
Fashion accessories,0.150000
Health and beauty,0.138235


In [36]:
df[df.City=='Mandalay']['Product line'].value_counts(normalize=True)

,proportion
Product line,
Fashion accessories,0.186747
Sports and travel,0.186747
Electronic accessories,0.165663
Health and beauty,0.159639
Food and beverages,0.150602
Home and lifestyle,0.150602


Як бачимо, в різних містах лідирують різні категорії товарів в продажах, тому перевіримо, чи в цьому полягає різниця в чеках, моджливо? Перевіримо вартість товарів (середню)в лідируючій позиції міста Mandalay, порівняємо іі з Yangon, а також загальну суму продажів Total лідируючої категорії і порівняємо її в двох містах

Очікуваний результат для кожного з міст:

In [56]:
df[(df['City'] == 'Yangon') & (df['Product line'] == 'Fashion accessories')]['Unit price'].mean()

np.float64(56.67039215686275)

In [57]:
df[(df['City'] == 'Mandalay') & (df['Product line'] == 'Fashion accessories')]['Unit price'].mean()

np.float64(54.84387096774193)

In [12]:
df[(df['City'] == 'Mandalay') & (df['Product line'] == 'Fashion accessories')]['Total'].sum()

np.float64(16413.3165)

In [13]:
df[(df['City'] == 'Yangon') & (df['Product line'] == 'Fashion accessories')]['Total'].sum()

np.float64(16332.5085)

Середня вартість товарів в саме цій категорії - якраз не відрізняється в обох містах, а от саме сума продажів загальна цієї категорії більша в місті Mandalay, що може давати різницю в Total. Але ця різнция все одно невелика.
Перевіримо в такому випадку розподіл Total тоді у містах і поглянемо на персентилі продажів.

In [14]:
df[df.City=='Yangon']['Total'].describe()


,Total
count,340.000000
mean,312.354031
std,231.640025
min,12.694500
25%,137.489625
50%,240.828000
75%,459.340875
max,1039.290000


In [15]:

df[df.City=='Mandalay']['Total'].describe()

,Total
count,332.000000
mean,319.872506
std,242.447565
min,18.637500
25%,119.936250
50%,252.882000
75%,465.585750
max,1022.490000


In [16]:
df[df.City=='Mandalay']['Total'].quantile(0.9)

np.float64(713.2681500000003)

In [17]:
df[df.City=='Yangon']['Total'].quantile(0.9)

np.float64(688.76115)

Можемо побачити і по 75 персентилю, і особливо по 90му, що в Mandalay чеки з більшими сумами, ніж у Yangon. Іншими словами, у Mandalay великі покупки є дорожчими.

Отже, розподіл чеків у Mandalay має більш виражений правий хвіст, що свідчить про більшу частку покупок з високою сумою. Ці великі чеки зміщують середнє значення Total вгору та пояснюють також вищий gross income у цьому місті.

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)

Підбиваючи підсумок, можна зробити такий висновок стосовно вищого загального доходу в місті Mandalay:
Він не має суттєвої різниці у порівнянні з містом Yangon, і викликаний, імовірніше за все, поєднанням двох факторів:
1) дещо вища вартість товарів в Mandalay порівняно з Yangon (unit price)
2) більша кільткість частки покупок з високою сумою,
Тобто верхня частина розподілу чеків (правий хвіст розподілу) є більш вираженою,  що зміщує середнє значення Total вгору. Можна скзаати,що в Mandalay є більша кількість паокупців, які роблять покупки на високий чек.
Водночас їх також не суттєва кількість, що не дає значної різниці в доходах між двома містами і є скоріше виключенням, ніж загальною тенденцією в різниці між цими двома містами.
Отже, можна скзаати, що продажі в цих двох містах приблизно на ожному рівні по багатьох показниах. Водночас, необхідно відслідковавути тендеенцію по покупках на великі чеки і дослідити можливість збільшити іхню кількість в місті Yangon.
